File: src/stores/tagTemplates.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { invoke } from '@tauri-apps/api/core'
import type { TagTemplate } from '@/types'

export const useTagTemplatesStore = defineStore('tagTemplates', () => {
  const templates = ref<TagTemplate[]>([])
  const loading = ref(false)
  const error = ref<string | null>(null)

  async function loadTemplates() {
    loading.value = true
    error.value = null
    try {
      templates.value = await invoke<TagTemplate[]>('get_tag_templates')
    } catch (e) {
      error.value = e as string
      console.error('Failed to load tag templates:', e)
    } finally {
      loading.value = false
    }
  }

  async function createTemplate(name: string, tagIds: number[]) {
    try {
      const id = await invoke<number>('create_tag_template', { name, tagIds })
      await loadTemplates()
      return id
    } catch (e) {
      error.value = e as string
      console.error('Failed to create tag template:', e)
      throw e
    }
  }

  async function applyTemplate(itemId: number, templateId: number) {
    try {
      await invoke('apply_tag_template', { itemId, templateId })
    } catch (e) {
      error.value = e as string
      console.error('Failed to apply tag template:', e)
      throw e
    }
  }

  async function deleteTemplate(id: number) {
    try {
      await invoke('delete_tag_template', { id })
      await loadTemplates()
    } catch (e) {
      error.value = e as string
      console.error('Failed to delete tag template:', e)
      throw e
    }
  }

  async function updateTemplate(id: number, name?: string, tagIds?: number[]) {
    try {
      await invoke('update_tag_template', { id, name, tagIds })
      await loadTemplates()
    } catch (e) {
      error.value = e as string
      console.error('Failed to update tag template:', e)
      throw e
    }
  }

  return {
    templates,
    loading,
    error,
    loadTemplates,
    createTemplate,
    applyTemplate,
    deleteTemplate,
    updateTemplate,
  }
})

File: src-tauri/src/commands/tag_templates.rs
//! TagTemplate Commands
//!
//! Thin adapters for tag template operations that delegate to TagTemplateService.

use crate::application::dto::{CreateTagTemplateDto, TagTemplateDto, UpdateTagTemplateDto};
use crate::error::{AppError, AppResult};
use crate::state::AppState;
use tauri::State;

#[tauri::command]
pub async fn create_tag_template(
    name: String,
    tag_ids: Vec<i64>,
    state: State<'_, AppState>,
) -> AppResult<i64> {
    let dto = CreateTagTemplateDto { name, tag_ids };

    state
        .tag_template_service
        .create(dto)
        .await
        .map_err(|e| AppError::InvalidInput(e.to_string()))
}

#[tauri::command]
pub async fn get_tag_templates(state: State<'_, AppState>) -> AppResult<Vec<TagTemplateDto>> {
    state
        .tag_template_service
        .get_all()
        .await
        .map_err(|e| AppError::InvalidInput(e.to_string()))
}

#[tauri::command]
pub async fn apply_tag_template(
    item_id: i64,
    template_id: i64,
    state: State<'_, AppState>,
) -> AppResult<()> {
    state
        .tag_template_service
        .apply_to_item(template_id, item_id)
        .await
        .map_err(|e| AppError::InvalidInput(e.to_string()))
}

#[tauri::command]
pub async fn delete_tag_template(id: i64, state: State<'_, AppState>) -> AppResult<()> {
    state
        .tag_template_service
        .delete(id)
        .await
        .map_err(|e| AppError::NotFound(e.to_string()))
}

#[tauri::command]
pub async fn update_tag_template(
    id: i64,
    name: Option<String>,
    tag_ids: Option<Vec<i64>>,
    state: State<'_, AppState>,
) -> AppResult<()> {
    let dto = UpdateTagTemplateDto { name, tag_ids };

    state
        .tag_template_service
        .update(id, dto)
        .await
        .map_err(|e| AppError::InvalidInput(e.to_string()))
}
